src/events/add-filter.js

```diff
@@ -1,17 +1,33 @@
 "use strict";
 
+var isFunction = require("../utils/is-function");
+
 var parseLogMessage = require("./parse-message/parse-log-message");
 
-function addFilter(blockStream, contractName, eventName, eventAbi, contracts, addSubscription, onMessage) {
-  if (!eventAbi) return false;
-  if (!eventAbi.contract || !eventAbi.signature || !eventAbi.inputs) return false;
-  if (!contracts[eventAbi.contract]) return false;
-  var contractAddress = contracts[eventAbi.contract];
-  addSubscription(contractAddress, eventAbi.signature, blockStream.addLogFilter({
+function addFilter(blockStream, contractName, callbacks, contractAbi, contracts, addSubscription) {
+  if (!contractAbi) return false;
+  if (!contracts[contractName]) return false;
+  var eventBySignature = {};
+  for (var eventName in callbacks) {
+    if (Object.prototype.hasOwnProperty.call(callbacks, eventName)) {
+      var eventAbi = contractAbi[eventName];
+      var callback = callbacks[eventName];
+      if (!isFunction(callback) || !eventAbi || !eventAbi.contract || !eventAbi.signature || !eventAbi.inputs) return false;
+      eventBySignature[eventAbi.signature] = {
+        eventName: eventName,
+        callback: callback,
+        abi: eventAbi.inputs,
+      };
+    }
+  }
+  var contractAddress = contracts[contractName];
+  addSubscription(contractAddress, blockStream.addLogFilter({
     address: contractAddress,
-    topics: [eventAbi.signature],
+    topics: [Object.keys(eventBySignature)],
   }), function (message) {
-    parseLogMessage(contractName, eventName, message, eventAbi.inputs, onMessage);
+    var event = eventBySignature[message.topics[0]];
+    if (!eventName) return console.warn("Ethereum Node provided data we did not request", contractAddress, message.topics[0]);
```

Comment:
It feels like the `parseLogMessage` line should be resilient to receiving calls that it doesn't understand.  If it is, then there is no need for this line.  If it isn't, then I worry a little about accidentally letting through a wrong log.

Reply:
If we're going to go to the single getLogs version, it seems like it might be more important to have a "hey, this matched but shouldn't have" warning?

Reply:
Yeah, I guess my question is really, "should that go here or down in `parseLogMessage` or deeper?  I tend to prefer having things be as resillient as possible, which usually ends up resulting in errors being handled pretty deep rather than high up.

I'm not familiar with this code, so certainly use your discretion as I may not be understanding the overarching patterns or intended usages.

Reply:
We can make parseLogMessage more resilient, too, but it does seem a bit awkward to enter it with undefined eventName, abi, and callback and let it notice. Especially since we are recognizing the risk of that happening by combining the 3 contracts into a single getLogs. It seems like we would be bothering the parser with the implementation of how we retrieved the logs